### PR TITLE
Add Evaluator to the pipeline

### PR DESCRIPTION
**Pull request by deep-diver**
This is verified to work on local and Vertex AI environments.

<p align="center">
  
</p>

Closes #30

### CODE REVIEW

File: training_pipeline/models/README.md [hunk 1]
```diff
@@ -2,7 +2,8 @@
...
   - `unet.py`: the model used in this project is a **U-NET** variant borrowed from [TensorFlow's official tutorial](https://www.tensorflow.org/tutorials/images/segmentation). the `build_model` function returns the U-NET implementation based on Keras, and it is used in the `train.py`.
-  - `signatures.py`: a signature let us add custom functions to a model when it is exported as `SavedModel`. This file defines a set of signatures for the model.
+  - `signatures.py`: a signature let us add custom functions to a model when it is exported as `SavedModel`. This file defines a set of signatures for the model. `model_exporter` defines a signature that handles requests when the model is deployed, `transform_features_signature` defines a signature that applies preprocessing logics from the `Transform` component to the input, and `tf_examples_serving_signature` defines a signature that gets predictions from the model after applying the preprocessing logics from the `Transform` component. 
+    - NOTE: difference between `model_exporter` and `tf_examples_serving_signature` is whether the preprocessing logics from the `Transform` component is applied.
```
Comment (sayakpaul) [inline training_pipeline/models/README.md:6]: Is there a flag within the utilities that controls this behaviour? If so, let's list them down.

File: training_pipeline/models/signatures.py [hunk 2]
```diff
@@ -66,3 +66,77 @@ def serving_fn(string_input):
...
+        get is the transformed data of the raw input.
+        """
+        feature_spec = tf_transform_output.raw_feature_spec()
+        parsed_features = tf.io.parse_example(serialized_tf_examples, feature_spec)
+
+        transformed_features = model.tft_layer(parsed_features)
+
+        return transformed_features
+
+    return serve_tf_examples_fn
```
Comment (sayakpaul) [inline training_pipeline/models/signatures.py:115]: This is a Swiss knife 🔪

File: training_pipeline/pipeline/configs.py [hunk 3]
```diff
@@ -42,6 +43,59 @@
...
+    we use the entire dataset to evaluate the model performance. If
+    you want to evaluate the model based on different slices of data
+    set, you should prepare TFRecords to have multiple features which
+    of each corresponds to each slices(or categories), then write the
+    slicing_specs options accordingly.
+"""
+EVAL_CONFIGS = tfma.EvalConfig(
+    model_specs=[
+        tfma.ModelSpec(
+            signature_name="from_examples",
```
Comment (sayakpaul) [inline training_pipeline/pipeline/configs.py:70]: How did you derive this name?

File: training_pipeline/pipeline/configs.py [hunk 4]
```diff
@@ -42,6 +43,59 @@
...
+            preprocessing_function_names=["transform_features"],
+            label_key="label_xf",
+            prediction_key="label_xf",
+        )
+    ],
+    slicing_specs=[tfma.SlicingSpec()],
```
Comment (sayakpaul) [inline training_pipeline/pipeline/configs.py:76]: Does it denote default behaviour? If so, what is the default behviour?

File: training_pipeline/pipeline/configs.py [hunk 5]
```diff
@@ -42,6 +43,59 @@
...
+    you want to evaluate the model based on different slices of data
+    set, you should prepare TFRecords to have multiple features which
+    of each corresponds to each slices(or categories), then write the
+    slicing_specs options accordingly.
+"""
+EVAL_CONFIGS = tfma.EvalConfig(
+    model_specs=[
+        tfma.ModelSpec(
+            signature_name="from_examples",
+            preprocessing_function_names=["transform_features"],
```
Comment (sayakpaul) [inline training_pipeline/pipeline/configs.py:71]: So, `transform_features` is a signature? If so, where is this signature coming from?

File: training_pipeline/pipeline/configs.py [hunk 6]
```diff
@@ -42,6 +43,59 @@
...
+    ],
+    slicing_specs=[tfma.SlicingSpec()],
+    metrics_specs=[
+        tfma.MetricsSpec(
+            metrics=[
+                tfma.MetricConfig(
+                    class_name="SparseCategoricalAccuracy",
+                    threshold=tfma.MetricThreshold(
+                        value_threshold=tfma.GenericValueThreshold(
+                            lower_bound={"value": 0.55}
```
Comment (sayakpaul) [inline training_pipeline/pipeline/configs.py:84]: Minimum accuracy for the Resolver component to derive it should graduate the current model to prod or not?

File: training_pipeline/pipeline/configs.py [hunk 7]
```diff
@@ -42,6 +43,59 @@
...
+                        ),
+                        # Change threshold will be ignored if there is no
+                        # baseline model resolved from MLMD (first run).
+                        change_threshold=tfma.GenericChangeThreshold(
+                            direction=tfma.MetricDirection.HIGHER_IS_BETTER,
+                            absolute={"value": -1e-3},
+                        ),
```
Comment (sayakpaul) [inline training_pipeline/pipeline/configs.py:91]: What do these denote?

**Review by sayakpaul — CHANGES_REQUESTED** [review]
There are a few changes here and there, but it's truly amazing to see how this turns out. 

Also, did you test this pipeline with two models having varying amount of accuracies?

File: training_pipeline/pipeline/configs.py (hunk 3 above)
Comment (deep-diver) [reply training_pipeline/pipeline/configs.py:70]: it is one of the signatures' name. I have searched the Internet for many materials, and I almost accidently copied the name from one of them. Please suggest a better name if possible!

File: training_pipeline/pipeline/configs.py (hunk 4 above)
Comment (deep-diver) [reply training_pipeline/pipeline/configs.py:76]: Actually, there is no default. If there is default, we don't have to specify things explicitly. However, default here means that it does not have special behaviours on `Slicing` that evaluates the model on the given dataset with one slice. 

each slice could mean different feature. in order to evaluate the model performance for different features differently, then TFRecords should hold the data accordingly. 
- for instance, instead of `image`, we could save different categories from the `image` into `street`, `building`, ....

File: training_pipeline/pipeline/configs.py (hunk 5 above)
Comment (deep-diver) [reply training_pipeline/pipeline/configs.py:71]: yes it is a signature, and as the name of `ModelSpec` suggests, it is from the model since model could have multiple signatures when it is saved/exported.

File: training_pipeline/pipeline/configs.py (hunk 6 above)
Comment (deep-diver) [reply training_pipeline/pipeline/configs.py:84]: yeah `tfma.GenericValueThreshold` is normally defined to set the minimum performance threshold. That means a model whose performance is better than this is going to be the first model, and it also means that it is only used when there is no model deployed in production yet.

File: training_pipeline/pipeline/configs.py (hunk 7 above)
Comment (deep-diver) [reply training_pipeline/pipeline/configs.py:91]: We can specify two models in the `Evaluator` component. One is the currently trained model, and the other one is the best model currently deployed(retrieved from the Artifact Store). 

`tfma.GenericChangeThreshold` let us to define the threshold by how much the currently trained model should be better than the previous model to replace it.
- we are evaluating the model based on `accuracy`, and higher score means better, so `tfma.MetricDirection.HIGHER_IS_BETTER` is set.

File: training_pipeline/models/README.md (hunk 1 above)
Comment (deep-diver) [reply training_pipeline/models/README.md:6]: sure!

**Comment by deep-diver** [comment]
> There are a few changes here and there, but it's truly amazing to see how this turns out.
> Also, did you test this pipeline with two models having varying amount of accuracies?

I am going to add more comments based on your questions! 

and, I have not tested with the two models scenario yet. we run the training for a couple of epochs, so the trained model could not achieve the baseline accuracy. with more epochs, we can get a good model, then change things (i.e. hyperparameters) will let us compare two models. 

I just checked that if everything is going fine without any errors. Gonna try it, but I guess that should be done in a separate PR for adjusting the metric values carefully.

**Comment by deep-diver** [comment]
@sayakpaul 

addressed your comments :)

File: training_pipeline/pipeline/configs.py (hunk 3 above)
Comment (sayakpaul) [reply training_pipeline/pipeline/configs.py:77]: `from_examples` doesn't tell me anything about the lineage of the artifacts. Could you comment on which component generates this signature?

File: training_pipeline/pipeline/configs.py (hunk 4 above)
Comment (sayakpaul) [reply training_pipeline/pipeline/configs.py:83]: Okay. Thanks for explaining. Very interesting.

So, the current configuration will only take the current eval dataset as is treating it to be a single slice?

File: training_pipeline/pipeline/configs.py (hunk 5 above)
Comment (sayakpaul) [reply training_pipeline/pipeline/configs.py:78]: Okay makes sense. This signature got added to the model in which component?

File: training_pipeline/pipeline/configs.py (hunk 6 above)
Comment (sayakpaul) [reply training_pipeline/pipeline/configs.py:96]: And which component tells the pipeline -- okay, I have a better model whose SparseCategoricalCrossentropy is better than the previously deployed model. `Resolver` I guess?

File: training_pipeline/pipeline/configs.py (hunk 7 above)
Comment (sayakpaul) [reply training_pipeline/pipeline/configs.py:107]: Makes a lot of sense. Thanks!

**Review by sayakpaul — APPROVED** [review]
> and, I have not tested with the two models scenario yet. we run the training for a couple of epochs, so the trained model could not achieve the baseline accuracy. with more epochs, we can get a good model, then change things (i.e. hyperparameters) will let us compare two models.

Sure! Maybe for sanity, we could lower the initial threshold and train two models for different numbers of epochs. That will let us verify the validity of the `Resolver` component faster.

But really amazing work overall. Feel free to merge the PR.

File: training_pipeline/pipeline/configs.py (hunk 4 above)
Comment (deep-diver) [reply training_pipeline/pipeline/configs.py:83]: Yup

File: training_pipeline/pipeline/configs.py (hunk 5 above)
Comment (deep-diver) [reply training_pipeline/pipeline/configs.py:78]: Trainer component

File: training_pipeline/pipeline/configs.py (hunk 6 above)
Comment (deep-diver) [reply training_pipeline/pipeline/configs.py:96]: Resolver is a kind of component to bring existing Artifact. So here it is used to look up previously trained best model
